### ISSUE TITLE
Add a MsgEnableFeeChannel to IBCFee (29-fee)

### ISSUE THREAD

**Issue by Reecepbcups** [cosmos/ibc-go#3622]
## Summary

IBCFee (29-fee) currently does not have a way to enable fees (fee_enabled_channels) on a channel during runtime. This  proposes a command be added to allow this from an authority address.

## Problem Definition

Currently there is no easy way to test IBCFee (29-fee) e2e or add new channels during run time without a binary upgrade. 

If you modify genesis to include IBCFee in E2E before the channel is created, the channel can never open (stuck in STATE_INIT forever). If you want to go the binary upgrade route, you have to maintain at minimum 2 binaries for every state breaking bump of a network (ex: base version image, image with ibcfee upgrade handler only)

I tried doing this here: https://github.com/CosmosContracts/juno/pull/684 but ran into many problems as its quite confusing to get setup.

## Proposal

Add a Msg (ex: MsgEnableFeeChannel) which only an authority can execute (gov module in line with how other SDK v47 modules do this) which runs `IBCFeeKeeper.SetFeeEnabled(ctx, PortId, ChannelId)`. 

This way a gov proposal (in production or testing) could add a new channel during runtime with a proposal by executing MsgEnableFeeChannel

(Note: Since SDKv47 allows for an array of messages tro gov, a singular ibcfeetype.MsgEnableFeeChannel is sufficient) 

____

#### For Admin Use

- [ ] Not duplicate issue
- [ ] Appropriate labels applied
- [ ] Appropriate contributors tagged/assigned

**Comment by Reecepbcups** [comment]
I would be happy to take this on if the team is interested. Would like to start contributing to ibc-go

**Comment by crodriguezvega** [comment]
Thanks for opening this issue, @Reecepbcups. I would like to ask some clarifying questions and add some comments.

> IBCFee (29-fee) currently does not have a way to enable fees (fee_enabled_channels) on a channel during runtime.

It is correct that at the moment it's not possible to enable fees on a channel that already exists. We are currently working on a solution ([channel upgradability](https://github.com/cosmos/ibc-go/issues/1599)) that will enable such use case. In a nutshell: if a channel already exists between two chains, the chains can perform a binary upgrade to include the fee middleware in the application stack and afterwards an upgrade can be initiated (in a permissioned or permissionless manner) to agree on a new channel version that includes the fee version besides the application version (this way during the upgrade handshake fees will be enabled on the channel).

> If you modify genesis to include IBCFee in E2E before the channel is created, the channel can never open (stuck in STATE_INIT forever)

I don't understand why the channel would get stuck in `INIT` state... If you specify a channel to be fee enabled in `genesis.json` all this does is to set a flag in state, but this flag does not have any use in the channel opening handshake. Am I missing something?

> If you want to go the binary upgrade route, you have to maintain at minimum 2 binaries for every state breaking bump of a network (ex: base version image, image with ibcfee upgrade handler only)

I also don't completely understand why you would need 2 binaries here. Could you please elaborate a bit more? Once a chain upgrades its binary and integrates the fee middleware in the application stack, then that should be it.

> Add a Msg (ex: MsgEnableFeeChannel) which only an authority can execute (gov module in line with how other SDK v47 modules do this) which runs IBCFeeKeeper.SetFeeEnabled(ctx, PortId, ChannelId).

If this message was added, you would still need a chain upgrade to integrate the fee middleware, otherwise setting the fee enabled flag on a channel will no have any effect. And even if the chain had already upgraded and integrated the fee middleware, the channel version would still only contain the application version and not the fee version, which would probably be problematic since it wouldn't be possible to be certain that both ends of the channel have agreed on the same version of the fee middleware (assuming that in the future there exists other versions of the fee middleware).

So do you see any reasons why the approach of using channel upgradability to enable "during runtime" fees in a channel would not be sufficient?

**Comment by Reecepbcups** [comment]
@crodriguezvega 

TLDR: I missed channel upgradability & misunderstood the 29-fee spec. Adding channel upgradability should properly resolve all the solutions I gave above in a simpler approach. Thank you for taking the time to explain this all to me.

_( I originally thought we could just attach a fee on ACK back to our chain, so even if the other channel did not have it implemented, it would still fund the relayer)_

---

> It is correct that at the moment it's not possible to enable fees on a channel that already exists. We are currently working on a solution (channel upgradability) that will enable such use case.

Ahh, I had missed this. Thank you

> I don't understand why the channel would get stuck in INIT state... If you specify a channel to be fee enabled in genesis.json all this does is to set a flag in state, but this flag does not have any use in the channel opening handshake. Am I missing something?

I do not think so. I got the genesis state directly from an export after chain upgrade. When I removed pre setting the value, IBC connections worked as expected. The only keeper which uses our IBCFee is wasm, but that should not modify channels being created. 

I had this same issue though with IBCFee when using it in the PacketFowardMiddleware on Juno (I use standard channel keeper now). So there may be something weird with IBCFee and creating new channels <- cc @agouin from our TG discussion the other week

> I also don't completely understand why you would need 2 binaries here.

For e2e testing, I would need the base binary. Software upgrade, new binary with the `keepers.IBCFeeKeeper.SetFeeEnabled(ctx, channel.PortId, channel.ChannelId)` logic. This seperate bin would need to include everything from the base branch, but the upgrade handler has to be in another bin due to software upgrades requiring the initial binary to not have the upgradeName loaded into the map.

With channel upgradability though this likely becomes a non issue / need anymore

**Comment by crodriguezvega** [comment]
> I had this same issue though with IBCFee when using it in the PacketFowardMiddleware on Juno (I use standard channel keeper now). So there may be something weird with IBCFee and creating new channel

Hmm, that's strange. If you're experiencing problems, please open a new issue and try to put as much details as possible with a reproduction scenario, so that we can have a look.

> For e2e testing, I would need the base binary. Software upgrade, new binary with the keepers.IBCFeeKeeper.SetFeeEnabled(ctx, channel.PortId, channel.ChannelId) logic. This seperate bin would need to include everything from the base branch, but the upgrade handler has to be in another bin due to software upgrades requiring the initial binary to not have the upgradeName loaded into the map.

But this flow is not different from any other upgrade flow, right? And once you do the upgrade, you don't need to keep any previous versions of the binary?
And doing `keepers.IBCFeeKeeper.SetFeeEnabled(ctx, channel.PortId, channel.ChannelId)` is not needed (or if you do it will not have the expected behaviour), since this will just set the flag on one side of the channel. If the counterparty doesn't have fee middleware integrated and enabled as well, then the incentivization would still not work. But even if you do a similar trick in the counterparty, the channel version will still be wrong (i.e. it will be still just `ics20-1` and not `{"fee_version":"ics29-1","app_version":"ics20-1"}`). To add fee middleware to your chain, you just need to [integrate it in the application stack](https://ibc.cosmos.network/main/middleware/ics29-fee/integration.html), and then if counterparties have also done the same, new channels can be created with incentivization. Once channel upgradability is released, then existing channels can also become incentivised.

> With channel upgradability though this likely becomes a non issue / need anymore

Then I will close this issue for now. :) Feel free still to comment here thought.
